实验六 Shell程序分析
=======

# 1. 实验类型

验证型，4学时，必做实验


# 2.  实验目的

掌握Shell程序的语法和程序结构，以及修改、调试的方法。

# 3. 题目描述

本次实验通过验证shell程序设计的基本要素，并以此修改和运行一段shell程序，从而掌握Shell程序查错、运行方法，为下次实验——设计Shell程序——打下基础。

# 4. 实验要求

* 基本层次要求：本次实验要求分析所给出程序的功能描述，并在程序中指定位置给出必要的注释；运行所给出的程序，通过查错分析，修改运行程序过程中所出现的错误。

* 扩展层次要求：利用结构程序设计的思想，使用程序流程图或N_S图描述程序的处理过程；总结Shell程序查错分析的方法。

# 5. 相关知识

结构程序设计是一种设计程序的技术，它采用自顶向下逐步求精的设计方法和单入口单出口的控制结构[10]。通过自顶向下逐步求精的方法，可以把一个复杂问题的解法分解和细化为一个由许多模块组成的层次结构的软件系统；单入口单出口的控制结构是指程序中不使用GO TO语句，以保证程序清晰流畅，易读易懂并且容易测试。结构程序设计采用三种基本的控制结构：顺序结构、选择（分支）结构和循环结构。

程序流程图和N_S图是软件详细设计阶段描述程序处理过程的工具，它们用来指明控制流程、处理功能、数据组织以及其它方面的实现细节，从而在编码阶段能把对设计的描述直接翻译成程序代码。程序流程图常用的符号如图6-1：

![程序流程图中常用的符号](images/6-1.png)
<center>图6-1 程序流程图中常用的符号</center>

用程序流程图描述的分支结构如图6-2表示：

![程序流程图描述的分支程序结构](images/6-2.png)
<center>图6-2 程序流程图描述的分支程序结构</center>

N_S图严格遵循了结构化程序设计思想，图6-3所示为N_S图描述的控制结构。

![N_S图描述的控制结构](images/6-3.png)
<center>图6-3 N_S图描述的控制结构</center>

BASH是Linux默认的Shell，程序为/bin/bash，其他的Shell如表6-1。本指导书中只介绍BASH，后文中若没有加上下文，则简称BASH环境中的Shell程序为Shell程序或者Shell脚本（Shell Script）。

<center>表6-1 Shell的历史</center>
![Shell的历史](images/6-4.png)

执行Shell脚本的方法有两种：一是将Shell脚本文件设置为可执行，直接执行该文件；二是指明使用bash解释、执行Shell脚本文件。

例如example.bash为Shell脚本文件，设置该文件为可执行后，在Shell下键入：

路径/example.bash

就可以执行该Shell脚本文件，因此，如果当前工作目录中包含example.bash，则使用命令“./example.bash”，其中“./”为径。

对Shell脚本程序进行调试，使用第二种执行方式：如执行example.bash文件，在Shell下键入：

bash 路径/example.bash

就可以执行该Shell脚本文件。其中，调试需要使用bash的选项：

-x

在变量替换之后但执行之前，显示命令文件的每一行。

-v

在执行之前显示命令文件的每一行，就像你输入一样。

-n

解释但不执行命令。

+x

关闭追踪。

因此，查错example.bash的方法为使用如下命令：

bash -x 路径/example.bash

例如：

In [1]:
cat workspace/example3

#!/bin/bash
# set user variale
year=2015
month=11
day=22
echo Today is $year-$month-$day


In [2]:
bash -x workspace/example3

+ year=2015
+ month=11
+ day=22
+ echo Today is 2015-11-22
Today is 2015-11-22


上述命令执行了exmple3脚本，由于使用了-x选项，该结果中包含了调试信息，即结果中以加号（+）开头的行，表示shell对该行命令的解释，比如“+ echo Today is 2015-11-22”一行表示shell在解释脚本中“echo Today is `$year-$month-$day`”一行时，将变量进行了替换再执行；没有以加号（+）开头的行则表示输出的信息。

### 实验任务1

运行/notebooks/workspace中的Shell脚本example.bash

# 6. 实验设备

实验者需要使用各自PC机上Windows操作系统的putty程序通过网络连接到Linux服务器（实验室内网为：10.11.8.17（暂时不可用），外网为：vlab.cs.swust.edu.cn），启动交互式实验指导书（浏览器方式，访问地址在启动程序中显示），并使用浏览器打开实验室内部的GitLab服务（[实验室内网访问](http://10.11.8.17:8081/)，[寝室外网访问](http://vlab.cs.swust.edu.cn:8081/)），学习和验证基本的实验操作，并完成实验指导书中的“实验任务”部分。

# 7. 实验指导

## 7.1、脚本程序基础

### (1) 多个命令

shell允许同一命令行输入多个命令运行，用分号隔开，比如：

In [3]:
date; uname -a

Mon Nov  7 03:17:36 UTC 2016
Linux 217d6088e0c3 4.4.0-45-generic #66-Ubuntu SMP Wed Oct 19 14:12:37 UTC 2016 x86_64 GNU/Linux


以上命令行也就是一个简单的脚本，执行了两个shell命令：date命令显示了当前日期和时间，uname显示了系统信息。

### 实验任务2

请使用一个命令行一次执行ls, pwd, whoami三个命令。

### (2) 脚本文件

使用命令行执行多个命令的方式有两个限制：命令行长度不能超过255个字符，而且每次执行需要将命令一次性输入，比较麻烦。可以通过创建一个文本文件，将需要执行的命令输入，然后运行这个文件，该文件就是shell脚本文件。

例如：创建一个文本文件test1，输入date和uname命令，该文件就称为shell脚本文件。

In [4]:
cat workspace/example1

#!/bin/bash
#This script displays the date and calendar.
date
uname -a


shell脚本中#开头的一行是注释，第一行#!称为魔数，用它来指示哪个shell要解释本脚本文件。

执行脚本文件有四种方式：

* bash 脚本文件名

* bash < 脚本文件名

* chmod u+x 脚本文件名; ./脚本文件名

* . 脚本文件名

例如：执行example1脚本：

In [5]:
bash workspace/example1

Mon Nov  7 03:18:14 UTC 2016
Linux 217d6088e0c3 4.4.0-45-generic #66-Ubuntu SMP Wed Oct 19 14:12:37 UTC 2016 x86_64 GNU/Linux


### 实验任务3

创建和执行一个脚本文件test1，该脚本文件执行ls, pwd, whoami三个命令。

### (3) 显示信息

echo命令可以显示一个文本字符串，比如：

In [6]:
echo "Hello, world!"

Hello, world!


通过echo命令可以在shell脚本中显示一些提示信息，比如：

In [7]:
cat workspace/example2

#!/bin/bash
#This script displays the date and calendar.
echo "This is date"
date
echo "This is system info"
uname -a


In [8]:
bash < workspace/example2

This is date
Mon Nov  7 03:18:52 UTC 2016
This is system info
Linux 217d6088e0c3 4.4.0-45-generic #66-Ubuntu SMP Wed Oct 19 14:12:37 UTC 2016 x86_64 GNU/Linux


上述命令使用了第二种执行脚本方法中的一种新的形式：使用输入重定向，即将脚本文件重定向为bash程序的输入（否则需要读入用户键盘的输入），从而执行了该脚本文件。（输入输出重定向的概念见后）

### (4) 使用变量

bash shell的变量包含两种：

* 环境变量

* 用户变量

bash shell使用环境变量来存储有关shell会话和工作环境的信息。这些信息包括系统的名称，登录到系统上的用户的名称，用户的默认主目录，shell查找程序的搜索路径等。使用set命令可以显示全部环境变量。如：

In [9]:
set

BASH=/bin/bash
BASHOPTS=checkwinsize:cmdhist:complete_fullquote:expand_aliases:extquote:force_fignore:histappend:hostcomplete:interactive_comments:progcomp:promptvars:sourcepath
BASH_ALIASES=()
BASH_ARGC=()
BASH_ARGV=()
BASH_CMDS=()
BASH_LINENO=()
BASH_SOURCE=()
BASH_VERSINFO=([0]="4" [1]="3" [2]="30" [3]="1" [4]="release" [5]="x86_64-pc-linux-gnu")
BASH_VERSION='4.3.30(1)-release'
COLUMNS=80
CONDA_DIR=/opt/conda
DEBIAN_FRONTEND=noninteractive
DIRSTACK=()
EUID=1002
GRANT_SUDO=yes
GROUPS=()
HISTCONTROL=ignoreboth
HISTFILE=/home/jovyan/.bash_history
HISTFILESIZE=2000
HISTSIZE=1000
HOME=/home/jovyan
HOSTNAME=217d6088e0c3
HOSTTYPE=x86_64
IFS=$' \t\n'
JPY_PARENT_PID=16
LANG=en_US.UTF-8
LANGUAGE=en_US.UTF-8
LC_ALL=en_US.UTF-8
LINES=24
LOGNAME=jovyan
LS_COLORS=
MACHTYPE=x86_64-pc-linux-gnu
MAIL=/var/mail/jovyan
MAILCHECK=60
NB_UID=1002
NB_USER=jovyan
OPTERR=1
OPTIND=1
OSTYPE=linux-gnu
PAGER=cat
PATH=/opt/conda/bin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PIP

使用“$变量名”的方式可以获取环境变量的值，比如：

In [10]:
echo $PATH

/opt/conda/bin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin


PATH环境变量是shell命令的查找路径，当执行shell命令出现“command not found”的错误时，需要考虑是命令不在该查找路径中，需要指明命令的路径。

通过“变量名=变量值”的方式可以定义一个用户变量。例如：

In [11]:
TODAY=2015-11-22

In [12]:
echo $TODAY

2015-11-22


如果在脚本中定义一个变量，则脚本执行结束后，该变量会被删除。例如：

In [13]:
cat workspace/example3

#!/bin/bash
# set user variale
year=2015
month=11
day=22
echo Today is $year-$month-$day


In [14]:
chmod u+x workspace/example3; workspace/example3

Today is 2015-11-22


In [15]:
echo $month

上述命令执行没有结果，是因为month变量在脚本中使用，该脚本执行结束后，变量就被删除了，即结果为空。若想使用脚本中定义的变量和函数等，需要使用如下执行方式：

In [16]:
. workspace/example3

Today is 2015-11-22


注意：命令行中，脚本前面是一个点，中间有空格。脚本执行后，看看month变量：

In [17]:
echo $month

11


可见，example3脚本中定义的变量month仍然有值。因此，这种执行脚本的方式通常用于在比较复杂的脚本程序中定义函数和变量。

反引号（\`，即美式键盘上“~”键，在“1”键旁边）可以将shell命令的输出赋值给变量，如：

In [18]:
DATE=`date`

In [19]:
echo $DATE

Mon Nov 7 03:31:17 UTC 2016


### 实验任务4

获取当前用户的默认主目录路径。

### (5) 输入输出重定向

重定向用于将某个文件用于输入或输出。其中，输出重定向是将命令的输出发送到一个文件中，输入重定向是将文件的内容输入给命令。例如：

In [20]:
date > workspace/out.txt

上述命令会将之前显示器上出现的命令输出保存到文件out.txt中。

In [21]:
cat workspace/out.txt

Mon Nov  7 03:31:26 UTC 2016


可见，out.txt文件中的内容的确是date命令的输出结果。若希望以添加方式输出到文件中，则使用双大于符号（>>），如：

In [22]:
echo "hello" >>workspace/out.txt

In [23]:
cat workspace/out.txt

Mon Nov  7 03:31:26 UTC 2016
hello


In [24]:
wc < workspace/out.txt

 2  7 35


wc命令提供了对数据中文本的技术，默认情况下，它会输出文本的行数、词数、字节数。上述命令行将文本文件重定向到wc命令，即对文件中的行数、词数、字节数进行了统计。

如果将输入重定向符号写成双小于号（<<），则称为内联输入重定向，需要指定一个文本标记作为输入数据的开始和结尾，而不是将文件作为输入。例如：

In [25]:
wc << EOF
test line1
test line2
test line3
EOF

 3  6 33


上述命令行是用wc命令统计EOF之间的文本的行数、词数和字节数。

### 实验任务5

试统计一下example1的行数、词数和字节数。

### (6) 管道

管道用作将一个命令的输出重定向到另外一个命令中：

command1 | command2

如：

In [26]:
dpkg -l | wc

    411    3981   47655


上述命令将dpkg -l命令的输出作为wc命令的输入，即统计了当前系统中已安装的包的数量。一般在命令行显示太多的时候，使用管道将结果重定向到more命令，以分页显示。例如使用：

dpkg -l | more

已分页显示所有已安装包的列表，请在访问[这里](/terminals/1)打开一个命令行窗口验证这个命令。

### 实验任务6

请使用管道，将dpkg -l命令显示的安装包的列表中vim相关的包过滤出来。

### (7) 数学运算

bash中进行数学运算有两种方式：

第一种方式是使用expr命令，命令行形式为：

expr 表达式

其中，表达式包括：

ARG1 + ARG2

加法操作

ARG1 - ARG2

减法操作

ARG1 * ARG2

乘法操作

ARG1 % ARG2

求余数操作

match STRING REGEXP

如果REGEXP匹配到了STRING，则返回真，否则，返回假。

例如，计算7\*8的值：

In [27]:
expr 7 \* 8

56


上述命令中的\符号为“转义符”，将其后的\*作为乘法，而不是shell命令行中的任意字符。注意表达式中的数字和操作符之间有空格

另外一种方法是使用方括号或双括号，用于将数学运算的结果赋值给某个变量，比如：

In [28]:
var1=$[1+5]

In [29]:
echo $var1

6


In [30]:
var2=$((2*19))

In [31]:
echo $var2

38


### 实验任务7

使用bash计算(1+5)\*19

### (8) 退出脚本

shell中每个命令执行结束，都会返回一个退出状态码，由特殊变量$?来保存，如：

In [32]:
date; echo $?

Mon Nov  7 03:35:01 UTC 2016
0


上述两个命令中，第一个date命令的退出状态码$?由第二个命令echo显示出来，状态码为0表示成功执行，其他表示执行有误。又如：

In [33]:
ddafdafa; echo $?

bash: ddafdafa: command not found
127


上述第一个命令ddafdafa命令不存在，所以执行有误，所以第二个命令echo显示其退出状态码$?为127（执行不成功）。

默认情况下，shell脚本的退出状态码由最后执行的命令来决定，如果想设置，则执行“exit 状态码”，其中，状态码最大不能操作255。例如：

In [34]:
cat workspace/example4

#!/bin/bash
echo "Hello, world!"


In [35]:
bash workspace/example4 ; echo $?

Hello, world!
0


In [36]:
cat workspace/example5

#!/bin/bash
echo "Hello, world!"
exit 100


In [37]:
bash workspace/example5;echo $?

Hello, world!
100


## 7.2、控制结构

### (1) 条件结构

条件结构用if then else命令来实现，其语法为：

`
if  测试命令
then  
  命令1
else
  命令2
fi
`

或：

`
if  测试命令
then
  命令1
elif  测试命令
then 
  命令2
fi
`

其中，测试命令为以下几种形式：

* “test 参数 值”的形式，如：     

test  -f  “text"

* “[ 参数 值 ]”的形式，如：

[  -f  “text"  ]

* "一般命令"的形式，如：

ls

注意：测试命令中"["和"]"与参数之间，以及参数与值之间，都有空格分割。否则会报语法错误。

test命令可以判断三类条件（测试方法），采用什么测试方法由“参数”部分来加以区别：

* 文件比较

参数为：
<left>表6-2 文件方面的测试</left>
![文件方面的测试](images/6-5.jpg)

* 字符串比较

参数为：
<left>表6-3 字符串方面的测试</left>
![字符串方面的测试](images/6-6.jpg)

* 数值比较

参数为：
<left>表6-4 数值方面的测试</left>
![数字方面的测试](images/6-7.jpg)

例如：

In [38]:
cat workspace/example6

#!/bin/bash
# disply home directory
if [ -d $HOME ]
then
  echo "Your HOME directory exists"
  cd $HOME
  ls -a
else
  echo "There is a problem with your HOME directory"
fi


In [39]:
bash workspace/example6

Your HOME directory exists
.   .bash_logout  .cache      .curlrc	.jupyter  .profile
..  .bashrc	  .continuum  .ipython	.local	  work


还可以使用case命令来实现多条件结构。命令行为：

`
case 字符串 in
    模式)
       语句
        ;;
    模式2 | 模式3)
       语句
       ;;
    *)
       默认执行的 语句
       ;;
esac
`

例如：

In [2]:
cat workspace/example12

#!/bin/bash
read -p "input a number:" NUMBER
case $NUMBER in
  1)
    echo "One"
    ;;
  2)
    echo "Two"
    ;;
  3)
    echo "Three"
    ;;
  4)
    echo "Four"
    ;;
  5)
    echo "Five"
    ;;
  6)
    echo "Six"
    ;;
  7)
    echo "Seven"
    ;;
  8)
    echo "Eight"
    ;;
  9)
    echo "Nine"
    ;;
  0)
    echo "Zero"
    ;;
  *)
    echo "Other number"
    ;;
esac


在命令窗口中执行该程序：

bash workspace/example12

输出结果为：

`
input a number:2 
Two
`

### 实验任务8

写一个bash脚本文件，功能为：提示输入一个用户名，判断该名字是否是当前操作者。（提示：输入用户名可使用read命令，可使用whoami命令获取当前操作者）

### (2) 循环结构

shell的循环结构由while命令、for命令、until命令实现。

for命令的格式为：

 `
for  变量  in  值表
do
    命令表
done
`

或： 

`
for ((e1;e2;e3))
do
    命令表
done
`

例如：

In [40]:
cat workspace/example7

#!/bin/bash
# for
for i in 0 1 2 3 4
do
  echo $i Hello, world!
done

for ((i=0;i<5;i++))
do
  echo $i Hello, world!
done


In [41]:
bash workspace/example7

0 Hello, world!
1 Hello, world!
2 Hello, world!
3 Hello, world!
4 Hello, world!
0 Hello, world!
1 Hello, world!
2 Hello, world!
3 Hello, world!
4 Hello, world!


while命令的格式为：

`
while  测试条件  
do
    命令表
done
`

例如：

In [42]:
cat workspace/example8

#!/bin/bash
# while
i=0
while [ $i -ne 5 ]
do
  echo $i Hello, world!
  i=$(($i+1))
done


In [43]:
bash workspace/example8

0 Hello, world!
1 Hello, world!
2 Hello, world!
3 Hello, world!
4 Hello, world!


until命令的格式为：

`
until   测试条件
do
    命令表
done
`

例如：

In [44]:
cat workspace/example9

#!/bin/bash
# while
i=0
until [ $i -eq 5 ]
do
  echo $i Hello, world!
  i=$(($i+1))
done


In [45]:
bash workspace/example9

0 Hello, world!
1 Hello, world!
2 Hello, world!
3 Hello, world!
4 Hello, world!


此外，控制循环的命令还包括：

* break命令

用于退出进行中的循环；

* continue命令

用于提早结束执行循环内部的命令，但不完全终止整个循环。

例如：

In [46]:
cat workspace/example10

#!/bin/bash
num=0
while [ $num -lt 5 ]
do
  num=$(($num+1))
  if [ $(($num%2)) -eq 0 ]
  then
    continue
  fi
echo $num
done


In [47]:
bash workspace/example10

1
3
5


### 实验任务9

写一个shell脚本，检查/notebooks/目录下的每个文件是一般文件还是目录。（提示：可使用for循环如：“for file in /notebooks/\*”来列举/notebooks/目录下的文件）

## 7.3、函数和参数传递

函数定义的格式为：

`
[function] 函数名()
{
命令行;
}
`

给函数传递参数使用特殊变量$1, $2......来进行，类似的，给脚本传递参数也是通过特殊变量$1, $2......来进行的。

函数内可以通过local定义局部变量，否则变量是针对整个脚本的全局变量。

例如：

In [48]:
cat workspace/example11

#!/bin/bash
# function
func()
{
local a="Test"
b="change"
echo "Let's begin now."
echo $a $b $c
echo $1 $2 $3
echo "The end."
return 0
}
a="Working directory"
b="is"
c=`pwd`
echo $a $b $c
func Welcome You, $1
echo $a $b $c


In [49]:
bash workspace/example11 me

Working directory is /notebooks
Let's begin now.
Test change /notebooks
Welcome You, me
The end.
Working directory change /notebooks


上述命令行使用bash workspace/example11 me方式启动example11脚本程序，其中me作为参数传递给shell。

在example11中，”func Welcome You, \$1“是对函数func的调用，参数是3个，分别为"Welcome","You,"以及变量"\$1"，其中，变量\$1是启动example11脚本时的参数"me"，在函数func的”echo \$1 \$2 \$3“语句中，\$1, \$2, \$3对应调用函数func的3个参数，分别是"Welcome","You,"以及变量"\$1"的值“me”。func函数中定义了局部变量a，所以该函数对a变量的修改没有影响到函数外的a变量，而函数中修改了变量b的值，因此函数外部的b变量也发生了变化。

## 8 实验任务

### 实验任务10

分析下面给出的Shell程序的功能，并运行、调试该程序（**注意：在putty的窗口中执行**）。

    #!/bin/bash
    IS_someone=`who | grep $1`
    while [ -z "$IS_someone" ]
    do
      sleep 3
      IS_someone=`who | grep $1`
    done
    echo "watch it,$1 is online!" | wall

### 实验任务11

请编写脚本，读入学号，输出学生成绩单中该学号对应的学生成绩（包括表的标题，表的名称），如：

![任务10-1](images/task10-1.png)
<center>图6-8 任务11-步骤1</center>

![任务10-2](images/task10-2.png)
<center>图6-9 任务11-步骤2</center>

成绩单文件为/notebooks/workspace/data.txt，内容为：

In [1]:
cat /notebooks/workspace/data.txt

课程总成绩						
学号	姓名	班级	实验课	平时成绩	期末考试	总成绩
20111001	白xx	信安1101	0.0 	0.0 	0 	0 
20111002	杜xx	信安1101	0.0 	0.0 	0 	0 
20111003	范xx	信安1101	0.0 	0.0 	0 	0 


### 实验任务12

画出实验启动脚本（实验包中的start.bash, config, function）的程序流程图（包括各函数的流程）。

# 9 实验思考

1、Shell脚本中的程序段有错误，但该程序段在运行中没有被该脚本执行，问错误能否被表现出来？

2、总结Shell脚本查错的基本方法和过程。